File: src/tutorial/part2/DemoObject.py
from m5.params import *
from m5.SimObject import SimObject


class DemoObject(SimObject):
    type = "DemoObject"
    cxx_header = "tutorial/part2/demo_object.hh"
    cxx_class = "gem5::DemoObject"

    time_to_wait = Param.Latency("Time before firing the event")
    number_of_fires = Param.Int(
        1, "Number of times to fire the event before goodbye"
    )
    farewell_object = Param.FarewellObject("A farewell object")


class FarewellObject(SimObject):
    type = "FarewellObject"
    cxx_header = "tutorial/part2/farewell_object.hh"
    cxx_class = "gem5::FarewellObject"

    buffer_size = Param.MemorySize(
        "1kB", "Size of buffer to fill with goodbye"
    )
    write_bandwidth = Param.MemoryBandwidth(
        "100MB/s", "Bandwidth to fill the buffer"
    )

File: src/tutorial/part2/demo_object.cc
#include "tutorial/part2/demo_object.hh"

#include <iostream>

#include "base/trace.hh"
#include "debug/Demo.hh"

namespace gem5
{

DemoObject::DemoObject(const DemoObjectParams &params) :
    SimObject(params),
    event([this]{processEvent();}, name()),
    farewell(params.farewell_object),
    myName(params.name),
    latency(params.time_to_wait),
    timesLeft(params.number_of_fires)
{
    DPRINTF(Demo, "DemoObject Created, with name %s\n", myName);
    panic_if(!farewell, "DemoObject must have a non-null FarewellObject");
}

void
DemoObject::processEvent()
{
    timesLeft--;
    DPRINTF(Demo, "Hello world! Processing the event! %d left\n", timesLeft);

    if (timesLeft <= 0) {
        DPRINTF(Demo, "Done firing!\n");
        farewell->sayGoodbye(myName);
    } else {
        schedule(event, curTick() + latency);
    }
}

void
DemoObject::startup()
{
    schedule(event, latency);
}

} // namespace gem5
